include/libopencm3/stm32/fdcan.h

```diff
@@ -0,0 +1,855 @@
+#define FDCAN_TXBTIE_TIE1				(1 << 1)
+#define FDCAN_TXBTIE_TIE2				(1 << 2)
+/**@}*/
+
+/** @defgroup fdcan_txbcie FDCAN_TXBCIE Transmit cancelled interrupt enable bits
+ * @{
+ */
+#define FDCAN_TXBCIE_CFIE0				(1 << 0)
+#define FDCAN_TXBCIE_CFIE1				(1 << 1)
+#define FDCAN_TXBCIE_CFIE2				(1 << 2)
+/**@}*/
+
+/* EFFL[2:0]: Event FIFO fill level*/
+#define FDCAN_TXEFS_EFFL_SHIFT			0
+#define FDCAN_TXEFS_EFFL_MASK			(__BITS(3) << FDCAN_TXEFS_EFFL_SHIFT)
+
+/* EFG[1:0]: Event FIFO get index */
+#define FDCAN_TXEFS_EFGI_SHIFT			8
+#define FDCAN_TXEFS_EFGI_MASK			(__BITS(2) << FDCAN_TXEFS_EFGI_SHIFT)
+
+/* EFPI[1:0]: Event FIFO put index */
+#define FDCAN_TXEFS_EFPI_SHIFT			16
+#define FDCAN_TXEFS_EFPI_MASK			(__BITS(2) << FDCAN_TXEFS_EFPI_SHIFT)
+
+#define FDCAN_TXEFS_EFF					(1 << 24)
+#define FDCAN_TXEFS_TEF					(1 << 25)
+
+/* EFAI[1:0]: Event FIFO acknowledge index */
+#define FDCAN_TXEFA_EFAI_SHIFT			0
+#define FDCAN_TXEFA_EFAI_MASK			(__BITS(2) << FDCAN_TXEFA_EFAI_SHIFT)
+
+
+/* PDIV[3:0]: Input clock divider */
+#define FDCAN_CKDIV_PDIV_SHIFT			0
+#define FDCAN_CKDIV_PDIV_MASK			(__BITS(4) << FDCAN_CKDIV_PDIV_SHIFT)
+
+/* --- FD-CAN memory block defines------------------------------------------ */
+
+/** Structure describing standard ID filter.
+ * Standard ID filter is composed of one 32bit value.
+ * This region of memory cannot be accessed in quantities less than 32bits.
+ */
+struct FDCAN_StandardFilter {
+	uint32_t type_id1_conf_id2; 		//<< aggregate of filter type, filter action and two IDs
+};
+
+#define FDCAN_SFT_SHIFT					30
+#define FDCAN_SFT_MASK					(__BITS(2) << FDCAN_SFT_SHIFT)
+
+/** @defgroup fdcan_sft Standard ID filter match type
+ * @{
+ */
+#define FDCAN_SFT_RANGE					(0x0 << FDCAN_SFT_SHIFT)	
+#define FDCAN_SFT_DUAL					(0x1 << FDCAN_SFT_SHIFT)
+#define FDCAN_SFT_ID_MASK				(0x2 << FDCAN_SFT_SHIFT)
+#define FDCAN_SFT_DISABLE				(0x3 << FDCAN_SFT_SHIFT)
+/**@}*/
+
+#define FDCAN_SFEC_SHIFT				27
+#define FDCAN_SFEC_MASK 				(__BITS(3) << FDCAN_SFEC_SHIFT)
+
+/** @defgroup fdcan_sfec Standard ID filter action
+ * @{
+ */
+#define FDCAN_SFEC_DISABLE 				(0x0 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_FIFO0				(0x1 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_FIFO1				(0x2 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_REJECT				(0x3 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_PRIO					(0x4 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_PRIO_FIFO0			(0x5 << FDCAN_SFEC_SHIFT)
+#define FDCAN_SFEC_PRIO_FIFO1			(0x6 << FDCAN_SFEC_SHIFT)
+/**@}*/
+
+/** Amount of standard filters allocated in Message RAM
+ * This number may vary between devices. 28 is value valid
+ * for STM32G4 
+ **/
+#define FDCAN_SFT_MAX_NR				28
+
+/* SFEC = 0x7 is unused */
+
+#define FDCAN_SFID1_SHIFT				16
+#define FDCAN_SFID1_MASK				(__BITS(11) << FDCAN_SFID1_SHIFT)
+
+#define FDCAN_SFID2_SHIFT				0
+#define FDCAN_SFID2_MASK				(__BITS(11) << FDCAN_SFID2_SHIFT)
+
+#define __CHANGE_BITS(tgt, value, mask) (tgt) = (tgt & ~(mask)) | (value & mask)
```

Comment:
none of this is emant to be used by users right?  put it in the c file.

if you were planning on having this be user API, please reconsider, we've never mixed macros and functions before, and it's not going to start now :)

Reply:
I think it is actually not even used as macros below are really a performance penalty if being used. And nobody sane enough would probably want to deal with FDCAN msgram on his own anyway. I will remove this & friends.